src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php

```diff
@@ -4,142 +4,183 @@
 
 namespace OpenTelemetry\SDK\Trace\SpanProcessor;
 
+use function assert;
+use Closure;
+use function count;
+use function intdiv;
 use InvalidArgumentException;
 use OpenTelemetry\Context\Context;
-use OpenTelemetry\SDK\Common\Environment\EnvironmentVariablesTrait;
-use OpenTelemetry\SDK\Common\Environment\Variables as Env;
 use OpenTelemetry\SDK\Common\Future\CancellationInterface;
-use OpenTelemetry\SDK\Common\Time\ClockFactory;
 use OpenTelemetry\SDK\Common\Time\ClockInterface;
-use OpenTelemetry\SDK\Common\Time\StopWatch;
-use OpenTelemetry\SDK\Common\Time\StopWatchFactory;
-use OpenTelemetry\SDK\Common\Time\Util as TimeUtil;
 use OpenTelemetry\SDK\Trace\ReadableSpanInterface;
 use OpenTelemetry\SDK\Trace\ReadWriteSpanInterface;
 use OpenTelemetry\SDK\Trace\SpanDataInterface;
 use OpenTelemetry\SDK\Trace\SpanExporterInterface;
 use OpenTelemetry\SDK\Trace\SpanProcessorInterface;
+use SplQueue;
 
 class BatchSpanProcessor implements SpanProcessorInterface
 {
-    use EnvironmentVariablesTrait;
-
     public const DEFAULT_SCHEDULE_DELAY = 5000;
     public const DEFAULT_EXPORT_TIMEOUT = 30000;
     public const DEFAULT_MAX_QUEUE_SIZE = 2048;
     public const DEFAULT_MAX_EXPORT_BATCH_SIZE = 512;
 
-    private ?SpanExporterInterface $exporter;
-    private ?int $maxQueueSize;
-    private ?int $scheduledDelayMillis;
-    // @todo: Please, check if this code is needed. It creates an error in phpstan, since it's not used
-    /** @phpstan-ignore-next-line */
-    private ?int $exporterTimeoutMillis;
-    private ?int $maxExportBatchSize;
-    private bool $running = true;
-    private StopWatch $stopwatch;
-
+    private SpanExporterInterface $exporter;
+    private ClockInterface $clock;
+    private int $maxQueueSize;
+    private int $maxBatchSize;
+    private int $scheduledDelayNanos;
+    private int $maxExportBatchSize;
+    private bool $autoFlush;
+
+    private ?int $nextScheduledRun = null;
+    private bool $running = false;
+    private int $batchId = 0;
     /** @var list<SpanDataInterface> */
-    private array $queue = [];
+    private array $batch = [];
+    /** @var SplQueue<list<SpanDataInterface>> */
+    private SplQueue $queue;
+    /** @var SplQueue<array{int, Closure}> */
+    private SplQueue $flush;
+
+    private bool $closed = false;
 
     public function __construct(
-        ?SpanExporterInterface $exporter,
-        ClockInterface $clock = null,
-        int $maxQueueSize = null,
-        int $scheduledDelayMillis = null,
-        int $exporterTimeoutMillis = null,
-        int $maxExportBatchSize = null
+        SpanExporterInterface $exporter,
+        ClockInterface $clock,
+        int $maxQueueSize = self::DEFAULT_MAX_QUEUE_SIZE,
+        int $scheduledDelayMillis = self::DEFAULT_SCHEDULE_DELAY,
+        int $exportTimeoutMillis = self::DEFAULT_EXPORT_TIMEOUT,
+        int $maxExportBatchSize = self::DEFAULT_MAX_EXPORT_BATCH_SIZE,
+        bool $autoFlush = true
```

Comment:
Should be changed to default `false` before beta to be consistent with async implementations that will only export once the event loop gains control. Using default `true` for now to keep the scope of the PR small / to not break examples etc.